app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,544 @@
+            bookOnDiskItem -> {
+              select(bookOnDiskItem);
+              return Unit.INSTANCE;
+            });
+    bookDelegate.setSelectionMode(SelectionMode.MULTI);
+    booksAdapter = new BooksOnDiskAdapter(bookDelegate
+        ,
+        BookOnDiskDelegate.LanguageDelegate.INSTANCE
+    );
+
+    presenter.attachView(this);
+
+    RecyclerView homeRecyclerView = findViewById(R.id.recycler_view_zim_host);
+    presenter.loadBooks();
+    homeRecyclerView.setAdapter(booksAdapter);
+
+    serviceConnection = new ServiceConnection() {
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        HotspotService.HotspotBinder binder = (HotspotService.HotspotBinder) service;
+        hotspotService = binder.getService();
+        bound = true;
+        hotspotService.registerCallBack(ZimHostActivity.this);
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName arg0) {
+        bound = false;
+      }
+    };
+
+    serviceIntent = new Intent(this, HotspotService.class);
+
+    startServerButton.setOnClickListener(v -> {
+      //Get the path of ZIMs user has selected
+      if (!isServerStarted) {
+        getSelectedBooksPath();
+        if (selectedBooksPath.size() > 0) {
+          startHotspotHelper();
+        } else {
+          Toast.makeText(ZimHostActivity.this, R.string.no_books_selected_toast_message,
+              Toast.LENGTH_SHORT).show();
+        }
+      } else {
+        startHotspotHelper();
+      }
+    });
+  }
+
+  private void startHotspotHelper() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      toggleHotspot();
+    } else {
+      //TO DO: show Dialog() + within that add check mobile Data check later.
+      //if (isMobileDataEnabled(context)) {
+      //  mobileDataDialog();
+      //} else {
+      if (isServerStarted) {
+        startService(ACTION_STOP_SERVER);
+      } else {
+        startHotspotDialog();
+      }
+      //}
+    }
+  }
+
+  private void getSelectedBooksPath() {
+    BooksOnDiskListItem.BookOnDisk bookOnDisk;
+
+    for (BooksOnDiskListItem item : booksAdapter.getItems()) {
+      if (item.isSelected()) {
+        bookOnDisk = (BooksOnDiskListItem.BookOnDisk) item;
+        File file = bookOnDisk.getFile();
+        selectedBooksPath.add(file.getAbsolutePath());
+        Log.v(TAG, "ZIM PATH : " + file.getAbsolutePath());
+      }
+    }
+  }
+
+  private void open(@Nullable BooksOnDiskListItem.BookOnDisk bookOnDisk) {
+    File file = bookOnDisk.getFile();
+    Intent zimFile = new Intent(this, MainActivity.class);
+    zimFile.setData(Uri.fromFile(file));
+    startActivity(zimFile);
+    finish();
+  }
+
+  private void select(@NonNull BooksOnDiskListItem.BookOnDisk bookOnDisk) {
+    ArrayList<BooksOnDiskListItem> booksList = new ArrayList<>();
+    for (BooksOnDiskListItem item : booksAdapter.getItems()) {
+      if (item.equals(bookOnDisk)) {
+        if (item.isSelected()) {
```

Comment:
So this is a common overcomplication where depending on the value of a boolean you then use a boolean. A simpler scenario that highlights this is 
```
if(myBoolean)
  return true
else
  return false
```
when all you actually need to say was
```
return myBoolean
```

much simpler and less work. So I'd like to see
```
item.setSelected(!item.isSelected())
```

Reply:
This might sound weird but the issue here was using `item.setSelected(!item.isSelected())` The files that user first checked and then unchecked were still in the `booksAdapter` as `isSelected =true`. So I had to break it down. Altho, it cracked me for a moment since both should work same but that wasn't the case.

Reply:
Fixed. 32128109804f352d74be139ac8a7d68b4f0eac2b